src/login/logind-session-dbus.c

```diff
@@ -421,6 +421,27 @@ static int method_set_display(sd_bus_message *message, void *userdata, sd_bus_er
         return sd_bus_reply_method_return(message, NULL);
 }
 
+static int method_set_tty(sd_bus_message *message, void *userdata, sd_bus_error *error) {
+        Session *s = ASSERT_PTR(userdata);
+        const char *tty;
+        int r;
+
+        assert(message);
+
+        r = sd_bus_message_read(message, "s", &tty);
+        if (r < 0)
+                return r;
+
```

Comment:
hmm, I think we probably should do some at least superficial validation of the argument and that it is actually a tty. ideally without opening it.

We probably shouldn't open the tty here though, hence isatty() (and the backing TCGETS ioctl) doesn't really work I guess.

i wonder if we can get away with a two-pronged approach:

1. validate proper device model ttys via sysfs, i.e. stat the thing, then use sd_device_new_from_stat_rdev() and sd_device_get_subsystem() to check the subsystem is "tty"
2. if that doesn't work, check if the device might be a pty, and for that use statfs, and then check the fs magic is DEVPTS_SUPER_MAGIC but the .st_rdev is not the one of /dev/pts/ptmx...

And two make this somewhat race-free, start by opening the inode via O_PATH and then only operate via fstat() and fstatfs() on that open fd.